src/swapd/runtime.rs

```diff
@@ -468,49 +476,56 @@ impl Runtime {
         }
     }
 
+    fn pending_requests(&mut self) -> &mut HashMap<ServiceId, Vec<PendingRequest>> {
+        &mut self.pending_requests.0
+    }
+
     fn continue_deferred_requests(
         &mut self,
         endpoints: &mut Endpoints,
         key: ServiceId,
         predicate: fn(&PendingRequest) -> bool,
     ) -> bool {
-        let mut success = true;
-        if let Some(pending_reqs) = self.pending_requests.remove(&key) {
-            let remaining_pending_reqs = pending_reqs
-                .into_iter()
-                .filter_map(|r| {
-                    if predicate(&r) {
-                        if let Ok(_) = match r.bus_id {
-                            ServiceBus::Ctl if r.dest == self.identity => {
-                                self.handle_rpc_ctl(endpoints, r.dest.clone(), r.request.clone())
-                            }
-                            ServiceBus::Msg if r.dest == self.identity => {
-                                self.handle_rpc_msg(endpoints, r.dest.clone(), r.request.clone())
+        let success = if let Some(pending_reqs) = self.pending_requests().remove(&key) {
+            let len0 = pending_reqs.len();
+            let remaining_pending_reqs: Vec<_> =
+                pending_reqs
+                    .into_iter()
+                    .filter_map(|r| {
+                        if predicate(&r) {
+                            if let Ok(_) = match &r.bus_id {
```

Comment:
You could get rid of this extra nesting level by converting the Err to Some(error) with `err()` and then mapping to `r` with `.map(|_| r)` :

```
handle_rpc....
.err()
.map(|_| r)
```

Reply:
> do you realize that Ok -> None, and Err -> Some?

Yes, the snippet I posted converts the Err(...) to Some(r) and the Ok(...) to None.